In [1]:
import pandas as pd
import numpy as np
import copy
import re

In [2]:
import pandas as pd
test = pd.read_csv("amazon_reviews_us_Kitchen_v1_00.tsv", sep = '\t',error_bad_lines = False)
test['label'] = -1
a1 = test.loc[test['star_rating']==1].sample(50000)
a2 = test.loc[test['star_rating']==2].sample(50000)
a3 = test.loc[test['star_rating']==3].sample(50000)
a4 = test.loc[test['star_rating']==4].sample(50000)
a5 = test.loc[test['star_rating']==5].sample(50000)


b'Skipping line 16148: expected 15 fields, saw 22\nSkipping line 20100: expected 15 fields, saw 22\nSkipping line 45178: expected 15 fields, saw 22\nSkipping line 48700: expected 15 fields, saw 22\nSkipping line 63331: expected 15 fields, saw 22\n'
b'Skipping line 86053: expected 15 fields, saw 22\nSkipping line 88858: expected 15 fields, saw 22\nSkipping line 115017: expected 15 fields, saw 22\n'
b'Skipping line 137366: expected 15 fields, saw 22\nSkipping line 139110: expected 15 fields, saw 22\nSkipping line 165540: expected 15 fields, saw 22\nSkipping line 171813: expected 15 fields, saw 22\n'
b'Skipping line 203723: expected 15 fields, saw 22\nSkipping line 209366: expected 15 fields, saw 22\nSkipping line 211310: expected 15 fields, saw 22\nSkipping line 246351: expected 15 fields, saw 22\nSkipping line 252364: expected 15 fields, saw 22\n'
b'Skipping line 267003: expected 15 fields, saw 22\nSkipping line 268957: expected 15 fields, saw 22\nSkipping line 303336: expected 15 field

b'Skipping line 1968846: expected 15 fields, saw 22\nSkipping line 1999941: expected 15 fields, saw 22\nSkipping line 2001492: expected 15 fields, saw 22\nSkipping line 2011204: expected 15 fields, saw 22\nSkipping line 2025295: expected 15 fields, saw 22\n'
b'Skipping line 2041266: expected 15 fields, saw 22\nSkipping line 2073314: expected 15 fields, saw 22\nSkipping line 2080133: expected 15 fields, saw 22\nSkipping line 2088521: expected 15 fields, saw 22\n'
b'Skipping line 2103490: expected 15 fields, saw 22\nSkipping line 2115278: expected 15 fields, saw 22\nSkipping line 2153174: expected 15 fields, saw 22\nSkipping line 2161731: expected 15 fields, saw 22\n'
b'Skipping line 2165250: expected 15 fields, saw 22\nSkipping line 2175132: expected 15 fields, saw 22\nSkipping line 2206817: expected 15 fields, saw 22\nSkipping line 2215848: expected 15 fields, saw 22\nSkipping line 2223811: expected 15 fields, saw 22\n'
b'Skipping line 2257265: expected 15 fields, saw 22\nSkipping line

In [3]:
new_test = pd.concat([a1,a2,a3,a4,a5])

In [4]:
new_test['lable'] = 1

In [5]:
new_test.label[test.star_rating>3] = 1
new_test.label[test.star_rating<3] = 2
new_test.label[test.star_rating==3] = 3
new_test = new_test[['label','review_body']] 

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so w

In [6]:
import gensim
model_1 = gensim.models.Word2Vec.load('save')

In [12]:
#data clearing
new_test['review_body'] = new_test['review_body'].str.lower()

def tag(x):
    return re.sub('<.*?>','',str(x))
new_test['review_body'] = new_test['review_body'].apply(lambda x:tag(x))

def url(x):
    return re.sub('(https?|ftp|file)://[-A-Za-z0-9+&@#/%?=~_|!:,.;]+[-A-Za-z0-9+&@#/%=~_|]','',str(x))

new_test['review_body'] = new_test['review_body'].apply(lambda x:url(x))

import contractions
new_test['review_body'] = new_test['review_body'].apply(lambda x:contractions.fix(x))

def non_alphabetical(x):
    return re.sub('[^a-zA-Z\s]','',str(x))
new_test['review_body'] = new_test['review_body'].apply(lambda x:non_alphabetical(x))
def extra_space(x):
    return re.sub( ' +',' ',str(x))
new_test['review_body'] = new_test['review_body'].apply(lambda x:extra_space(x))

In [14]:
#Pre-processing

import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
stop_words_set = set(stopwords.words('english'))
from nltk import word_tokenize, pos_tag

def stop_words(x):
    word_tokens = word_tokenize(x)
    temp = []
    for i in word_tokens:
        if i not in stop_words_set:
            temp.append(i)
    return temp
new_test['review_body'] = new_test['review_body'].apply(lambda x:stop_words(x))

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/chenlin/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [15]:
new_test

label                                        review_body
1261010      2           [victorinox, product, first, word, read]
149938       2  [match, size, color, tho, color, name, also, e...
4379464      2  [advertised, add, char, broil, cooker, purchas...
28297        2                                             [yuck]
4844336      2  [bought, toaster, black, decker, name, sorry, ...
...        ...                                                ...
3962262      1  [well, made, lid, stays, tight, else, ask, mug...
1501108      1    [love, dishes, added, coffee, cups, set, years]
2533959      1  [pug, lover, like, love, cool, fridge, magnet,...
48365        1  [bought, pan, make, cake, yo, celebrate, offic...
4145163      1  [person, gave, toaster, oven, bad, review, sai...

[250000 rows x 2 columns]

In [16]:
def vector_avg(x):
    i = 0
    n = 0
    a = np.zeros(300)
    while i<len(x):
        if x[i] in model_1.wv:
            a += model_1.wv[x[i]]
        else:
            n += 1
        i = i+1
    if n == 0: return a
    else:return a/(len(x)-n)

In [17]:
new_test['avg_vector'] = new_test['review_body'].apply(lambda x:vector_avg(x))

In [32]:
new_test['y'] = -1

In [34]:
new_test['y'][new_test['label'] == 1] =0

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [35]:
new_test['y'][new_test['label'] == 2] =1

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [36]:
new_test['y'][new_test['label'] == 3] =2

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [37]:
new_test

label                                        review_body  \
1261010      2           [victorinox, product, first, word, read]   
149938       2  [match, size, color, tho, color, name, also, e...   
4379464      2  [advertised, add, char, broil, cooker, purchas...   
28297        2                                             [yuck]   
4844336      2  [bought, toaster, black, decker, name, sorry, ...   
...        ...                                                ...   
3962262      1  [well, made, lid, stays, tight, else, ask, mug...   
1501108      1    [love, dishes, added, coffee, cups, set, years]   
2533959      1  [pug, lover, like, love, cool, fridge, magnet,...   
48365        1  [bought, pan, make, cake, yo, celebrate, offic...   
4145163      1  [person, gave, toaster, oven, bad, review, sai...   

                                                avg_vector  y  
1261010  [0.3361993785947561, 0.3755444842390716, 0.205...  1  
149938   [0.023530922830104828, 0.7736800406128168, 0.0...  1  
4379464  [0.9732713364064693, 1.8002999844029546, 0.767...  1  
28297    [-0.30345696210861206, 0.005608719307929277, -...  1  
4844336  [0.05253013025219666, 0.0007164585744825805, 0...  1  
...                                                    ... ..  
3962262  [0.6963680814951658, 1.8281475193798542, -0.74...  0  
1501108  [1.1396409068256617, 0.5796800553798676, 0.723...  0  
2533959  [0.031026512011885644, 0.06857175541420778, -0...  0  
48365    [0.04685636181134863, 0.04049028242713705, -0....  0  
4145163  [0.06293770963466673, 0.027035569575536152, -0...  0  

[250000 rows x 4 columns]

In [42]:
X = new_test['avg_vector']
Y = new_test['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

# ternary

In [43]:
# import libraries
import torch
from torch.utils.data import DataLoader, Dataset
import torchvision
import torchvision.transforms as transforms
from torch.utils.data.sampler import SubsetRandomSampler
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

In [44]:
from torch.utils.data import DataLoader, Dataset
from torch.utils.data import TensorDataset, DataLoader

In [45]:

# number of subprocesses to use for data loading
num_workers = 0
# how many samples per batch to load
batch_size = 20
# percentage of training set to use as validation
#valid_size = 0.2

DatasetTrain = TensorDataset(torch.from_numpy(x_train),torch.from_numpy(y_train))
        
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

train_loader=torch.utils.data.DataLoader(DatasetTrain,batch_size=batch_size,shuffle=True,drop_last=True, num_workers=0)

valid_loader=torch.utils.data.DataLoader(DatasetTest, batch_size=batch_size, drop_last=True,num_workers=0)

In [46]:
for xi ,yi in train_loader:
    print(xi,yi)
    break

tensor([[ 2.4691e+00,  2.8362e+00,  6.5291e-01,  ...,  1.4724e+00,
          4.6802e+00, -3.3048e+00],
        [ 8.3865e-01,  1.8521e-01,  1.7748e-01,  ...,  1.5910e+00,
          1.3063e+00, -8.6692e-01],
        [-2.9935e-01,  5.9200e-01,  2.7225e-01,  ...,  7.0387e-01,
          5.6117e-01,  5.3711e-01],
        ...,
        [ 7.1137e-01,  1.1082e+00,  9.6256e-01,  ...,  1.1243e+00,
          1.8219e+00, -1.9473e-02],
        [-9.6564e-01,  2.0634e+00,  4.8056e-01,  ...,  1.2328e+00,
          1.0309e+00, -1.7420e+00],
        [ 3.0430e-02,  6.6379e-02,  4.6200e-03,  ...,  1.9586e-02,
          7.6102e-02, -6.7758e-02]], dtype=torch.float64) tensor([1, 0, 1, 0, 0, 1, 2, 2, 1, 2, 1, 2, 1, 1, 2, 2, 1, 1, 1, 1])


In [47]:
import torch.nn as nn
import torch.nn.functional as F

# define the NN architecture
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        # number of hidden nodes in each layer 
        hidden_1 = 50
        hidden_2 = 10
        # linear layer (300 -> hidden_1)
        self.fc1 = nn.Linear(300, hidden_1)
        # linear layer (n_hidden -> hidden_2)
        self.fc2 = nn.Linear(hidden_1, hidden_2)
        # linear layer (n_hidden -> 10)
        self.fc3 = nn.Linear(hidden_2, 3)
        # dropout layer (p=0.2)
        # dropout prevents overfitting of data
        self.dropout = nn.Dropout(0.2)

    def forward(self, x):
        # flatten image input
        x = x.view(-1, 300)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        x = self.dropout(x)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc2(x))
        # add dropout layer
        x = self.dropout(x)
        # add output layer
        x = self.fc3(x)
        return x

# initialize the NN
model = Net()
print(model)

Net(
  (fc1): Linear(in_features=300, out_features=50, bias=True)
  (fc2): Linear(in_features=50, out_features=10, bias=True)
  (fc3): Linear(in_features=10, out_features=3, bias=True)
  (dropout): Dropout(p=0.2, inplace=False)
)


In [48]:
# import libraries
import torch
## Specify loss and optimization functions

# specify loss function
criterion = nn.CrossEntropyLoss()

# specify optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

In [50]:
# number of epochs to train the model
n_epochs = 20  # suggest training between 20-50 epochs

model.train() # prep model for training

for epoch in range(n_epochs):
    # monitor training loss
    train_loss = 0.0
    
    ###################
    # train the model #
    ###################
    for data, target in train_loader:
        # clear the gradients of all optimized variables
        optimizer.zero_grad()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = model(data.float())
        # calculate the loss
        loss = criterion(output, target)
        # backward pass: compute gradient of the loss with respect to model parameters
        loss.backward()
        # perform a single optimization step (parameter update)
        optimizer.step()
        # update running training loss
        train_loss += loss.item()*data.size(0)
        
    # print training statistics 
    # calculate average loss over an epoch
    train_loss = train_loss/len(train_loader.dataset)

    print('Epoch: {} \tTraining Loss: {:.6f}'.format(
        epoch+1, 
        train_loss
        ))

Epoch: 1 	Training Loss: 0.830280
Epoch: 2 	Training Loss: 0.778383
Epoch: 3 	Training Loss: 0.760423
Epoch: 4 	Training Loss: 0.752586
Epoch: 5 	Training Loss: 0.747508
Epoch: 6 	Training Loss: 0.743031
Epoch: 7 	Training Loss: 0.739228
Epoch: 8 	Training Loss: 0.735913
Epoch: 9 	Training Loss: 0.733647
Epoch: 10 	Training Loss: 0.731600
Epoch: 11 	Training Loss: 0.729493
Epoch: 12 	Training Loss: 0.727222
Epoch: 13 	Training Loss: 0.725659
Epoch: 14 	Training Loss: 0.726512
Epoch: 15 	Training Loss: 0.723097
Epoch: 16 	Training Loss: 0.722527
Epoch: 17 	Training Loss: 0.721406
Epoch: 18 	Training Loss: 0.719671
Epoch: 19 	Training Loss: 0.717970
Epoch: 20 	Training Loss: 0.717478


In [51]:
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

total_num = 0
correct = 0
for x_temp,y_real in DatasetTest:
    res = model(x_temp.float())
    _, pred = torch.max(res,1)
    total_num += 1
    if int(pred) == int(y_real):
        correct += 1

In [52]:
accuracy_mymodel_avg =  correct/total_num

In [56]:
print('The accuracy for ternary label from my model using average vector is ', accuracy_mymodel_avg)

The accuracy for ternary label from my model using average vector is  0.6855


In [54]:
def vector_concat(x):
    i = 0
    n_x = 0
    length = len(x)
    if length == 0:
        return np.zeros(3000)
    a = np.zeros(300)
    
    while i<10:
        if n_x < length:
            if x[n_x] in model_1.wv :
                if i == 0:
                    ans = copy.deepcopy(model_1.wv[x[n_x]])
                    i += 1
                else:
                    ans = np.concatenate((ans,model_1.wv[x[n_x]]), axis=None)
                    i += 1
            n_x += 1
        else:
            if i == 0:
                ans = a
                i += 1
            else:
                ans = np.concatenate((ans,a), axis=None)
                i += 1
    return ans

In [57]:
new_test['concat_vec'] = new_test['review_body'].apply(lambda x:vector_concat(x))

In [63]:
new_test

label                                        review_body  \
1261010      2           [victorinox, product, first, word, read]   
149938       2  [match, size, color, tho, color, name, also, e...   
4379464      2  [advertised, add, char, broil, cooker, purchas...   
28297        2                                             [yuck]   
4844336      2  [bought, toaster, black, decker, name, sorry, ...   
...        ...                                                ...   
3962262      1  [well, made, lid, stays, tight, else, ask, mug...   
1501108      1    [love, dishes, added, coffee, cups, set, years]   
2533959      1  [pug, lover, like, love, cool, fridge, magnet,...   
48365        1  [bought, pan, make, cake, yo, celebrate, offic...   
4145163      1  [person, gave, toaster, oven, bad, review, sai...   

                                                avg_vector  y  \
1261010  [0.3361993785947561, 0.3755444842390716, 0.205...  1   
149938   [0.023530922830104828, 0.7736800406128168, 0.0...  1   
4379464  [0.9732713364064693, 1.8002999844029546, 0.767...  1   
28297    [-0.30345696210861206, 0.005608719307929277, -...  1   
4844336  [0.05253013025219666, 0.0007164585744825805, 0...  1   
...                                                    ... ..   
3962262  [0.6963680814951658, 1.8281475193798542, -0.74...  0   
1501108  [1.1396409068256617, 0.5796800553798676, 0.723...  0   
2533959  [0.031026512011885644, 0.06857175541420778, -0...  0   
48365    [0.04685636181134863, 0.04049028242713705, -0....  0   
4145163  [0.06293770963466673, 0.027035569575536152, -0...  0   

                                                concat_vec  
1261010  [0.11119204014539719, -0.04113255813717842, 0....  
149938   [-0.15528874, 0.082478374, 0.16331409, 0.11810...  
4379464  [0.0612764, 0.18021023, -0.078093216, -0.21873...  
28297    [-0.30345696210861206, 0.005608719307929277, -...  
4844336  [0.06608908, -0.18412489, 0.31485054, -0.13780...  
...                                                    ...  
3962262  [-0.023777505, 0.19637382, 0.094618395, 0.0448...  
1501108  [0.22806349396705627, 0.019244728609919548, 0....  
2533959  [0.41755006, 0.32110247, 0.17386933, 0.1215716...  
48365    [0.06608908, -0.18412489, 0.31485054, -0.13780...  
4145163  [0.16218725, 0.10329889, -0.3650812, -0.031553...  

[250000 rows x 5 columns]

In [65]:
X = new_test['concat_vec']
Y = new_test['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

In [68]:
len(x_train[0])

3000

In [66]:

# number of subprocesses to use for data loading
num_workers = 0
# how many samples per batch to load
batch_size = 20
# percentage of training set to use as validation
#valid_size = 0.2

DatasetTrain = TensorDataset(torch.from_numpy(x_train),torch.from_numpy(y_train))
        
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

train_loader=torch.utils.data.DataLoader(DatasetTrain,batch_size=batch_size,shuffle=True,drop_last=True, num_workers=0)

valid_loader=torch.utils.data.DataLoader(DatasetTest, batch_size=batch_size, drop_last=True,num_workers=0)

In [69]:
import torch.nn as nn
import torch.nn.functional as F

# define the NN architecture
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        # number of hidden nodes in each layer 
        hidden_1 = 50
        hidden_2 = 10
        # linear layer (300 -> hidden_1)
        self.fc1 = nn.Linear(3000, hidden_1)
        # linear layer (n_hidden -> hidden_2)
        self.fc2 = nn.Linear(hidden_1, hidden_2)
        # linear layer (n_hidden -> 10)
        self.fc3 = nn.Linear(hidden_2, 3)
        # dropout layer (p=0.2)
        # dropout prevents overfitting of data
        self.dropout = nn.Dropout(0.2)

    def forward(self, x):
        # flatten image input
        x = x.view(-1, 3000)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        x = self.dropout(x)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc2(x))
        # add dropout layer
        x = self.dropout(x)
        # add output layer
        x = self.fc3(x)
        return x

# initialize the NN
model = Net()
print(model)

Net(
  (fc1): Linear(in_features=3000, out_features=50, bias=True)
  (fc2): Linear(in_features=50, out_features=10, bias=True)
  (fc3): Linear(in_features=10, out_features=3, bias=True)
  (dropout): Dropout(p=0.2, inplace=False)
)


In [70]:
# import libraries
import torch
## Specify loss and optimization functions

# specify loss function
criterion = nn.CrossEntropyLoss()

# specify optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

In [71]:
# number of epochs to train the model
n_epochs = 20  # suggest training between 20-50 epochs

model.train() # prep model for training

for epoch in range(n_epochs):
    # monitor training loss
    train_loss = 0.0
    
    ###################
    # train the model #
    ###################
    for data, target in train_loader:
        # clear the gradients of all optimized variables
        optimizer.zero_grad()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = model(data.float())
        # calculate the loss
        loss = criterion(output, target)
        # backward pass: compute gradient of the loss with respect to model parameters
        loss.backward()
        # perform a single optimization step (parameter update)
        optimizer.step()
        # update running training loss
        train_loss += loss.item()*data.size(0)
        
    # print training statistics 
    # calculate average loss over an epoch
    train_loss = train_loss/len(train_loader.dataset)

    print('Epoch: {} \tTraining Loss: {:.6f}'.format(
        epoch+1, 
        train_loss
        ))

Epoch: 1 	Training Loss: 0.919045
Epoch: 2 	Training Loss: 0.838901
Epoch: 3 	Training Loss: 0.816031
Epoch: 4 	Training Loss: 0.799537
Epoch: 5 	Training Loss: 0.785014
Epoch: 6 	Training Loss: 0.770246
Epoch: 7 	Training Loss: 0.757134
Epoch: 8 	Training Loss: 0.744034
Epoch: 9 	Training Loss: 0.730689
Epoch: 10 	Training Loss: 0.717832
Epoch: 11 	Training Loss: 0.704430
Epoch: 12 	Training Loss: 0.692234
Epoch: 13 	Training Loss: 0.681463
Epoch: 14 	Training Loss: 0.670074
Epoch: 15 	Training Loss: 0.660044
Epoch: 16 	Training Loss: 0.648021
Epoch: 17 	Training Loss: 0.640992
Epoch: 18 	Training Loss: 0.630216
Epoch: 19 	Training Loss: 0.623205
Epoch: 20 	Training Loss: 0.614358


In [73]:
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

total_num = 0
correct = 0
for x_temp,y_real in DatasetTest:
    res = model(x_temp.float())
    _, pred = torch.max(res,1)
    total_num += 1
    if int(pred) == int(y_real):
        correct += 1

In [74]:
accuracy_mymodel_concat_ternary =  correct/total_num

In [75]:
print('The accuracy for ternary label from my model using concat vector is ', accuracy_mymodel_concat_ternary)

The accuracy for ternary label from my model using concat vector is  0.61106


# google news 300 for ternary label

# average vector

In [77]:
import gensim.downloader as api
wv = api.load('word2vec-google-news-300')


In [81]:
googel_300 = new_test[['label','review_body']]

In [83]:
def vector_concat(x):
    i = 0
    n_x = 0
    length = len(x)
    if length == 0:
        return np.zeros(3000)
    a = np.zeros(300)
    
    while i<10:
        if n_x < length:
            if x[n_x] in wv :
                if i == 0:
                    ans = wv[x[n_x]]
                    i += 1
                else:
                    ans = np.concatenate((ans,wv[x[n_x]]), axis=None)
                    i += 1
            n_x += 1
        else:
            if i == 0:
                ans = a
                i += 1
            else:
                ans = np.concatenate((ans,a), axis=None)
                i += 1
    return ans

In [84]:
googel_300['concat_vec'] = googel_300['review_body'].apply(lambda x:vector_concat(x))

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [85]:
def vector_avg(x):
    i = 0
    n = 0
    a = np.zeros(300)
    while i<len(x):
        if x[i] in wv:
            a += wv[x[i]]
        else:
            n += 1
        i = i+1
    if n == 0: return a
    else:return a/(len(x)-n)

In [86]:
googel_300['avg_vector'] = googel_300['review_body'].apply(lambda x:vector_avg(x))

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [96]:
googel_300['y'] = -1
googel_300['y'] [googel_300['label']== 1] = 0
googel_300['y'] [googel_300['label']== 2] = 1
googel_300['y'] [googel_300['label']== 3] = 2

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing t

In [97]:
googel_300

label                                        review_body  \
1261010      2           [victorinox, product, first, word, read]   
149938       2  [match, size, color, tho, color, name, also, e...   
4379464      2  [advertised, add, char, broil, cooker, purchas...   
28297        2                                             [yuck]   
4844336      2  [bought, toaster, black, decker, name, sorry, ...   
...        ...                                                ...   
3962262      1  [well, made, lid, stays, tight, else, ask, mug...   
1501108      1    [love, dishes, added, coffee, cups, set, years]   
2533959      1  [pug, lover, like, love, cool, fridge, magnet,...   
48365        1  [bought, pan, make, cake, yo, celebrate, offic...   
4145163      1  [person, gave, toaster, oven, bad, review, sai...   

                                                concat_vec  \
1261010  [-0.0615234375, 0.09521484375, 0.1337890625, 0...   
149938   [-0.15527344, 0.025024414, 0.064941406, -0.124...   
4379464  [-0.044433594, -0.017456055, -0.10986328, -0.1...   
28297    [-0.09765625, 0.1728515625, 0.25, 0.3515625, -...   
4844336  [0.16699219, -0.05419922, -0.087402344, 0.0196...   
...                                                    ...   
3962262  [-0.08251953, 0.022460938, -0.14941406, 0.0991...   
1501108  [0.10302734375, -0.15234375, 0.02587890625, 0....   
2533959  [-0.07128906, -0.013244629, -0.16503906, 0.197...   
48365    [0.16699219, -0.05419922, -0.087402344, 0.0196...   
4145163  [0.27539062, -0.24707031, 0.017211914, 0.16796...   

                                                avg_vector  y  
1261010  [0.09716796875, -0.015625, 0.033599853515625, ...  1  
149938   [-0.073455810546875, 0.9627685546875, -0.23001...  1  
4379464  [0.16021728515625, 0.9654693603515625, -0.6411...  1  
28297    [-0.09765625, 0.1728515625, 0.25, 0.3515625, -...  1  
4844336  [3.5324554443359375, 0.277191162109375, 2.2111...  1  
...                                                    ... ..  
3962262  [0.4085693359375, 1.667144775390625, -0.440200...  0  
1501108  [-0.83380126953125, 0.60302734375, -0.04724121...  0  
2533959  [0.7264404296875, 0.86065673828125, -0.7277069...  0  
48365    [-0.018688746861049106, 0.016660417829241072, ...  0  
4145163  [-0.03807124385127315, 0.061185483579282406, 0...  0  

[250000 rows x 5 columns]

In [98]:
X = googel_300['avg_vector']
Y = googel_300['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

In [99]:

# number of subprocesses to use for data loading
num_workers = 0
# how many samples per batch to load
batch_size = 20
# percentage of training set to use as validation
#valid_size = 0.2

DatasetTrain = TensorDataset(torch.from_numpy(x_train),torch.from_numpy(y_train))
        
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

train_loader=torch.utils.data.DataLoader(DatasetTrain,batch_size=batch_size,shuffle=True,drop_last=True, num_workers=0)

valid_loader=torch.utils.data.DataLoader(DatasetTest, batch_size=batch_size, drop_last=True,num_workers=0)

In [100]:
import torch.nn as nn
import torch.nn.functional as F

# define the NN architecture
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        # number of hidden nodes in each layer 
        hidden_1 = 50
        hidden_2 = 10
        # linear layer (300 -> hidden_1)
        self.fc1 = nn.Linear(300, hidden_1)
        # linear layer (n_hidden -> hidden_2)
        self.fc2 = nn.Linear(hidden_1, hidden_2)
        # linear layer (n_hidden -> 10)
        self.fc3 = nn.Linear(hidden_2, 3)
        # dropout layer (p=0.2)
        # dropout prevents overfitting of data
        self.dropout = nn.Dropout(0.2)

    def forward(self, x):
        # flatten image input
        x = x.view(-1, 300)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        x = self.dropout(x)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc2(x))
        # add dropout layer
        x = self.dropout(x)
        # add output layer
        x = self.fc3(x)
        return x

# initialize the NN
model = Net()
print(model)


Net(
  (fc1): Linear(in_features=300, out_features=50, bias=True)
  (fc2): Linear(in_features=50, out_features=10, bias=True)
  (fc3): Linear(in_features=10, out_features=3, bias=True)
  (dropout): Dropout(p=0.2, inplace=False)
)


In [101]:

# specify loss function
criterion = nn.CrossEntropyLoss()

# specify optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

In [102]:
# number of epochs to train the model
n_epochs = 20  # suggest training between 20-50 epochs

model.train() # prep model for training

for epoch in range(n_epochs):
    # monitor training loss
    train_loss = 0.0
    
    ###################
    # train the model #
    ###################
    for data, target in train_loader:
        # clear the gradients of all optimized variables
        optimizer.zero_grad()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = model(data.float())
        # calculate the loss
        loss = criterion(output, target)
        # backward pass: compute gradient of the loss with respect to model parameters
        loss.backward()
        # perform a single optimization step (parameter update)
        optimizer.step()
        # update running training loss
        train_loss += loss.item()*data.size(0)
        
    # print training statistics 
    # calculate average loss over an epoch
    train_loss = train_loss/len(train_loader.dataset)

    print('Epoch: {} \tTraining Loss: {:.6f}'.format(
        epoch+1, 
        train_loss
        ))

Epoch: 1 	Training Loss: 0.888198
Epoch: 2 	Training Loss: 0.819735
Epoch: 3 	Training Loss: 0.803842
Epoch: 4 	Training Loss: 0.794879
Epoch: 5 	Training Loss: 0.789147
Epoch: 6 	Training Loss: 0.784515
Epoch: 7 	Training Loss: 0.780191
Epoch: 8 	Training Loss: 0.778002
Epoch: 9 	Training Loss: 0.774452
Epoch: 10 	Training Loss: 0.772460
Epoch: 11 	Training Loss: 0.769765
Epoch: 12 	Training Loss: 0.767997
Epoch: 13 	Training Loss: 0.765558
Epoch: 14 	Training Loss: 0.765464
Epoch: 15 	Training Loss: 0.762639
Epoch: 16 	Training Loss: 0.760900
Epoch: 17 	Training Loss: 0.759239
Epoch: 18 	Training Loss: 0.757018
Epoch: 19 	Training Loss: 0.755700
Epoch: 20 	Training Loss: 0.754198


In [103]:
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

total_num = 0
correct = 0
for x_temp,y_real in DatasetTest:
    res = model(x_temp.float())
    _, pred = torch.max(res,1)
    total_num += 1
    if int(pred) == int(y_real):
        correct += 1

In [104]:
accuracy_googel_avg_ternary =  correct/total_num

In [105]:
print('The accuracy for ternary label from googel 300 model using average vector is ', accuracy_googel_avg_ternary)

The accuracy for ternary label from googel 300 model using average vector is  0.66946


# google news 300 for ternary label 


# concat vecter

In [107]:
X = googel_300['concat_vec']
Y = googel_300['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

In [108]:

# number of subprocesses to use for data loading
num_workers = 0
# how many samples per batch to load
batch_size = 20
# percentage of training set to use as validation
#valid_size = 0.2

DatasetTrain = TensorDataset(torch.from_numpy(x_train),torch.from_numpy(y_train))
        
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

train_loader=torch.utils.data.DataLoader(DatasetTrain,batch_size=batch_size,shuffle=True,drop_last=True, num_workers=0)

valid_loader=torch.utils.data.DataLoader(DatasetTest, batch_size=batch_size, drop_last=True,num_workers=0)

In [109]:

# define the NN architecture
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        # number of hidden nodes in each layer 
        hidden_1 = 50
        hidden_2 = 10
        # linear layer (300 -> hidden_1)
        self.fc1 = nn.Linear(3000, hidden_1)
        # linear layer (n_hidden -> hidden_2)
        self.fc2 = nn.Linear(hidden_1, hidden_2)
        # linear layer (n_hidden -> 10)
        self.fc3 = nn.Linear(hidden_2, 3)
        # dropout layer (p=0.2)
        # dropout prevents overfitting of data
        self.dropout = nn.Dropout(0.2)

    def forward(self, x):
        # flatten image input
        x = x.view(-1, 3000)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        x = self.dropout(x)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc2(x))
        # add dropout layer
        x = self.dropout(x)
        # add output layer
        x = self.fc3(x)
        return x

# initialize the NN
model = Net()
print(model)


Net(
  (fc1): Linear(in_features=3000, out_features=50, bias=True)
  (fc2): Linear(in_features=50, out_features=10, bias=True)
  (fc3): Linear(in_features=10, out_features=3, bias=True)
  (dropout): Dropout(p=0.2, inplace=False)
)


In [110]:

# specify loss function
criterion = nn.CrossEntropyLoss()

# specify optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

In [111]:
# number of epochs to train the model
n_epochs = 20  # suggest training between 20-50 epochs

model.train() # prep model for training

for epoch in range(n_epochs):
    # monitor training loss
    train_loss = 0.0
    
    ###################
    # train the model #
    ###################
    for data, target in train_loader:
        # clear the gradients of all optimized variables
        optimizer.zero_grad()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = model(data.float())
        # calculate the loss
        loss = criterion(output, target)
        # backward pass: compute gradient of the loss with respect to model parameters
        loss.backward()
        # perform a single optimization step (parameter update)
        optimizer.step()
        # update running training loss
        train_loss += loss.item()*data.size(0)
        
    # print training statistics 
    # calculate average loss over an epoch
    train_loss = train_loss/len(train_loader.dataset)

    print('Epoch: {} \tTraining Loss: {:.6f}'.format(
        epoch+1, 
        train_loss
        ))

Epoch: 1 	Training Loss: 0.949798
Epoch: 2 	Training Loss: 0.872999
Epoch: 3 	Training Loss: 0.850158
Epoch: 4 	Training Loss: 0.833774
Epoch: 5 	Training Loss: 0.818439
Epoch: 6 	Training Loss: 0.805019
Epoch: 7 	Training Loss: 0.789864
Epoch: 8 	Training Loss: 0.773915
Epoch: 9 	Training Loss: 0.758746
Epoch: 10 	Training Loss: 0.745903
Epoch: 11 	Training Loss: 0.731850
Epoch: 12 	Training Loss: 0.718900
Epoch: 13 	Training Loss: 0.705711
Epoch: 14 	Training Loss: 0.692805
Epoch: 15 	Training Loss: 0.682410
Epoch: 16 	Training Loss: 0.672709
Epoch: 17 	Training Loss: 0.661839
Epoch: 18 	Training Loss: 0.652909
Epoch: 19 	Training Loss: 0.642177
Epoch: 20 	Training Loss: 0.636413


In [122]:

total_num = 0
correct = 0
for x_temp,y_real in DatasetTest:
    res = model(x_temp.float())
    _, pred = torch.max(res,1)
    total_num += 1
    if int(pred) == int(y_real):
        correct += 1

In [123]:
accuracy_google_concat_ternary =  correct/total_num

In [124]:
print('The accuracy for ternary label from google 300 model using concat vector is ', accuracy_google_concat_ternary)

The accuracy for ternary label from google 300 model using concat vector is  0.59616


# google news 300 for binary label

# average vector

In [126]:
googel_300_binary = googel_300.loc[googel_300['label']!=3]

In [133]:
X = googel_300_binary['avg_vector']
Y = googel_300_binary['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

In [134]:
# number of subprocesses to use for data loading
num_workers = 0
# how many samples per batch to load
batch_size = 20
# percentage of training set to use as validation
#valid_size = 0.2

DatasetTrain = TensorDataset(torch.from_numpy(x_train),torch.from_numpy(y_train))
        
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

train_loader=torch.utils.data.DataLoader(DatasetTrain,batch_size=batch_size,shuffle=True,drop_last=True, num_workers=0)

valid_loader=torch.utils.data.DataLoader(DatasetTest, batch_size=batch_size, drop_last=True,num_workers=0)

In [135]:

# define the NN architecture
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        # number of hidden nodes in each layer 
        hidden_1 = 50
        hidden_2 = 10
        # linear layer (300 -> hidden_1)
        self.fc1 = nn.Linear(300, hidden_1)
        # linear layer (n_hidden -> hidden_2)
        self.fc2 = nn.Linear(hidden_1, hidden_2)
        # linear layer (n_hidden -> 10)
        self.fc3 = nn.Linear(hidden_2, 2)
        # dropout layer (p=0.2)
        # dropout prevents overfitting of data
        self.dropout = nn.Dropout(0.2)

    def forward(self, x):
        # flatten image input
        x = x.view(-1, 300)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        x = self.dropout(x)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc2(x))
        # add dropout layer
        x = self.dropout(x)
        # add output layer
        x = self.fc3(x)
        return x

# initialize the NN
model = Net()
print(model)


Net(
  (fc1): Linear(in_features=300, out_features=50, bias=True)
  (fc2): Linear(in_features=50, out_features=10, bias=True)
  (fc3): Linear(in_features=10, out_features=2, bias=True)
  (dropout): Dropout(p=0.2, inplace=False)
)


In [136]:

# specify loss function
criterion = nn.CrossEntropyLoss()

# specify optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

In [137]:
# number of epochs to train the model
n_epochs = 20  # suggest training between 20-50 epochs

model.train() # prep model for training

for epoch in range(n_epochs):
    # monitor training loss
    train_loss = 0.0
    
    ###################
    # train the model #
    ###################
    for data, target in train_loader:
        # clear the gradients of all optimized variables
        optimizer.zero_grad()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = model(data.float())
        # calculate the loss
        loss = criterion(output, target)
        # backward pass: compute gradient of the loss with respect to model parameters
        loss.backward()
        # perform a single optimization step (parameter update)
        optimizer.step()
        # update running training loss
        train_loss += loss.item()*data.size(0)
        
    # print training statistics 
    # calculate average loss over an epoch
    train_loss = train_loss/len(train_loader.dataset)

    print('Epoch: {} \tTraining Loss: {:.6f}'.format(
        epoch+1, 
        train_loss
        ))

Epoch: 1 	Training Loss: 0.516591
Epoch: 2 	Training Loss: 0.436113
Epoch: 3 	Training Loss: 0.418995
Epoch: 4 	Training Loss: 0.408833
Epoch: 5 	Training Loss: 0.401627
Epoch: 6 	Training Loss: 0.398218
Epoch: 7 	Training Loss: 0.394187
Epoch: 8 	Training Loss: 0.390971
Epoch: 9 	Training Loss: 0.386998
Epoch: 10 	Training Loss: 0.383374
Epoch: 11 	Training Loss: 0.381110
Epoch: 12 	Training Loss: 0.378476
Epoch: 13 	Training Loss: 0.376614
Epoch: 14 	Training Loss: 0.374721
Epoch: 15 	Training Loss: 0.373161
Epoch: 16 	Training Loss: 0.370029
Epoch: 17 	Training Loss: 0.369939
Epoch: 18 	Training Loss: 0.366963
Epoch: 19 	Training Loss: 0.366842
Epoch: 20 	Training Loss: 0.364092


In [138]:

total_num = 0
correct = 0
for x_temp,y_real in DatasetTest:
    res = model(x_temp.float())
    _, pred = torch.max(res,1)
    total_num += 1
    if int(pred) == int(y_real):
        correct += 1

In [139]:
accuracy_googel_avg_binary =  correct/total_num

In [140]:
print('The accuracy for binary label from google 300 model using avg vector is ', accuracy_googel_avg_binary)

The accuracy for binary label from google 300 model using avg vector is  0.831475


# google news 300 for binary label
concat vector

# concat vector

In [142]:
X = googel_300_binary['concat_vec']
Y = googel_300_binary['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

In [143]:
# number of subprocesses to use for data loading
num_workers = 0
# how many samples per batch to load
batch_size = 20
# percentage of training set to use as validation
#valid_size = 0.2

DatasetTrain = TensorDataset(torch.from_numpy(x_train),torch.from_numpy(y_train))
        
DatasetTest=TensorDataset(torch.from_numpy(x_test),torch.from_numpy(y_test))

train_loader=torch.utils.data.DataLoader(DatasetTrain,batch_size=batch_size,shuffle=True,drop_last=True, num_workers=0)

valid_loader=torch.utils.data.DataLoader(DatasetTest, batch_size=batch_size, drop_last=True,num_workers=0)

In [144]:

# define the NN architecture
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        # number of hidden nodes in each layer 
        hidden_1 = 50
        hidden_2 = 10
        # linear layer (300 -> hidden_1)
        self.fc1 = nn.Linear(3000, hidden_1)
        # linear layer (n_hidden -> hidden_2)
        self.fc2 = nn.Linear(hidden_1, hidden_2)
        # linear layer (n_hidden -> 10)
        self.fc3 = nn.Linear(hidden_2, 2)
        # dropout layer (p=0.2)
        # dropout prevents overfitting of data
        self.dropout = nn.Dropout(0.2)

    def forward(self, x):
        # flatten image input
        x = x.view(-1, 3000)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc1(x))
        # add dropout layer
        x = self.dropout(x)
        # add hidden layer, with relu activation function
        x = F.relu(self.fc2(x))
        # add dropout layer
        x = self.dropout(x)
        # add output layer
        x = self.fc3(x)
        return x

# initialize the NN
model = Net()
print(model)


Net(
  (fc1): Linear(in_features=3000, out_features=50, bias=True)
  (fc2): Linear(in_features=50, out_features=10, bias=True)
  (fc3): Linear(in_features=10, out_features=2, bias=True)
  (dropout): Dropout(p=0.2, inplace=False)
)


In [145]:

# specify loss function
criterion = nn.CrossEntropyLoss()

# specify optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

In [146]:
# number of epochs to train the model
n_epochs = 20  # suggest training between 20-50 epochs

model.train() # prep model for training

for epoch in range(n_epochs):
    # monitor training loss
    train_loss = 0.0
    
    ###################
    # train the model #
    ###################
    for data, target in train_loader:
        # clear the gradients of all optimized variables
        optimizer.zero_grad()
        # forward pass: compute predicted outputs by passing inputs to the model
        output = model(data.float())
        # calculate the loss
        loss = criterion(output, target)
        # backward pass: compute gradient of the loss with respect to model parameters
        loss.backward()
        # perform a single optimization step (parameter update)
        optimizer.step()
        # update running training loss
        train_loss += loss.item()*data.size(0)
        
    # print training statistics 
    # calculate average loss over an epoch
    train_loss = train_loss/len(train_loader.dataset)

    print('Epoch: {} \tTraining Loss: {:.6f}'.format(
        epoch+1, 
        train_loss
        ))

Epoch: 1 	Training Loss: 0.555557
Epoch: 2 	Training Loss: 0.485170
Epoch: 3 	Training Loss: 0.467148
Epoch: 4 	Training Loss: 0.450466
Epoch: 5 	Training Loss: 0.433529
Epoch: 6 	Training Loss: 0.415610
Epoch: 7 	Training Loss: 0.398995
Epoch: 8 	Training Loss: 0.381185
Epoch: 9 	Training Loss: 0.362999
Epoch: 10 	Training Loss: 0.346290
Epoch: 11 	Training Loss: 0.331080
Epoch: 12 	Training Loss: 0.314752
Epoch: 13 	Training Loss: 0.301338
Epoch: 14 	Training Loss: 0.288892
Epoch: 15 	Training Loss: 0.276443
Epoch: 16 	Training Loss: 0.264201
Epoch: 17 	Training Loss: 0.254010
Epoch: 18 	Training Loss: 0.243546
Epoch: 19 	Training Loss: 0.236213
Epoch: 20 	Training Loss: 0.229487


In [147]:
total_num = 0
correct = 0
for x_temp,y_real in DatasetTest:
    res = model(x_temp.float())
    _, pred = torch.max(res,1)
    total_num += 1
    if int(pred) == int(y_real):
        correct += 1

In [148]:
accuracy_googel_concat_binary =  correct/total_num

In [149]:
print('The accuracy for binary label from google 300 model using concat vector is ', accuracy_googel_avg_binary)

The accuracy for binary label from google 300 model using concat vector is  0.831475


## conclusion
The accuracy for binary label from google 300 model using avg vector is  0.831475

The accuracy for binary label from google 300 model using concat vector is  0.831475

The accuracy for ternary label from google 300 model using concat vector is  0.59616

The accuracy for ternary label from googel 300 model using average vector is  0.66946

The accuracy for ternary label from my model using concat vector is  0.61106

The accuracy for ternary label from my model using average vector is  0.6855